Automattic/woocommerce-payments
# Minimal Payments API Client

A first draft on addressing #10. I hope this is roughly what the issue had in mind.

Some points for discussion:

* WordPress itself, and a lot of plugins have quite flat directory structures. I've found it useful to have a more nested structure here, but happy to change for consistency.
* Some of the data structures in the requests and responses are likely to get quite complex. In order to manage this I've opted to accept and return objects in favour of raw arrays. As well as documenting the potential complexity it makes handling things like mandatory properties or optional properties with defaults easier (as well as providing hints in IDEs etc). Some requests we could probably get away with simply passing arguments instead of a request object. I've stuck with an object for now but this can be changed.
* I've type hinted what I can given our PHP version (so object instance type hinting but no scalar hinting). I find it helps debugging if things fail early on, but I haven't seen this done in many places - I can remove the hinting if there's a reason for that.
* I've copied the approach taken by the WooCommerce Services plugin for creating service objects and injecting their dependencies (we just have the one service with a single dependency for now).
* The new code written should meet the coding standards. The only exception is the file level comments. I find that when you're doing a class per file these are a bit much. Is there anything sensible we can put up there that doesn't duplicate what's already in the class description? Perhaps moving the main summary into the file level comment and having a more basic comment on the class?
* The client throws basic `Exception` exceptions, we should probably throw something more specific and have a think about what sort of exception hierarchy we want.
* We have some classes prefixed `wc-payment` and some `wc-payments`. What's the thinking here? I may need to rename some classes to match.

Changes from some other payment gateways:

* I have made the API class non-static to aid testing and wiring up dependencies
* Each request has a public function, rather than having a common request function that handles lots of different calls. The reasoning here is similar to what we've been discussing for the server API.

Outstanding for this PR:

* [x] Some unit tests - ~~so not completely sure everything is working as it should just yet~~

Things that will need doing later (shout if you would rather see them in this PR):

* User agent reporting / sending HTTP headers
* Idempotency keys
* Logging
* Various filters (although there is an argument for simply adding these as and when we need them?)
* Handle failed payments in `process_payment`

## Testing this PR

* Run `./bin/install-wp-tests.sh <test_database_name> <db_user_name> <db_password> <db_host>:<db_port>`
* Run `composer test` to run the unit tests
* Run `composer phpcs .` to check coding standards

These things will be checked by Travis, but doesn't hurt to check locally as well.
